[ad_integration] Use LDAPS for Simple AD integration test

tilne opened the pull request:
* Explicitly set file permissions being tested
*  Use NLB to implement LDAPS for SimpleAD
* Create integ test directory in private subnets
* Use better resource names in AD stack
* Manage existence of ~/.ssh directories when testing automatic keypair generation
* Reset SSH connections after cluster update
* Run AD integration integ test on ubuntu20

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Use better resource names in AD stack

Previously the names auto-generated by CDK were used.

Signed-off-by: Tim Lane <tilne@amazon.com>

Commit:
Create integ test directory in private subnets

Signed-off-by: Tim Lane <tilne@amazon.com>

Commit:
Use NLB to implement LDAPS for SimpleAD

Signed-off-by: Tim Lane <tilne@amazon.com>

Commit:
Reset SSH connections after cluster update

Signed-off-by: Tim Lane <tilne@amazon.com>

Commit:
Run AD integration integ test on ubuntu20

Signed-off-by: Tim Lane <tilne@amazon.com>

Commit:
Manage existence of ~/.ssh directories when testing automatic keypair generation

Signed-off-by: Tim Lane <tilne@amazon.com>

hanwen-pcluste commented on tests/integration-tests/tests/ad_integration/test_ad_integration.py:
```diff
@@ -46,17 +47,20 @@ def get_infra_stack_outputs(stack_name):
     """Return a dict used to set values for config file parameters."""
-    infra_stack_outputs = get_infra_stack_outputs(stack_name)
+    stack_outputs = {
+        stack_type: get_infra_stack_outputs(stack_name)
+        for stack_type, stack_name in [("directory", directory_stack_name), ("nlb", nlb_stack_name)]
```
minor: why using a loop instead of laying out the dict directly? Or maybe just use two variables `directory_stack_outputs` and `nlb_stack_outputs`?

tilne replied:
i don't have a great reason. i think the idea was to use one line instead of two, but i agree looking at it now that it would probably be simpler to read if i used two variables instead. changing it now.

tilne replied:
fixed.

hanwen-pcluste commented on tests/integration-tests/tests/ad_integration/test_ad_integration.py:
```diff
@@ -155,7 +159,9 @@ def _store_secret(secret):
 @pytest.fixture(scope="module")
 def directory_factory(request, cfn_stacks_factory):
-    created_directory_stacks = {}
+    # TODO: do we need to worry about regions?
```
Why wouldn't we care about the region?

tilne replied:
even within the same region clusters won't these tests probably be running in their own processes, and thus not be able to share this data structure?

hanwen-pcluste replied:
Yes. If the tests are in different processes, they couldn't share the AD.
The idea was to reduce the creation time of AD and share it within each process.
Sharing it across processes is not easy (e.g. we need to handle racing condition)

tilne replied:
yeah agreed. some sort of lock would be needed. the reason i leave the TODO asking if we need to handle regions is because i think we can assume that tests running within the same process will be in the same region. but it doesn't hurt anything to handle it, and it makes it safer. i'll remove the region TODO.

tilne replied:
done

hanwen-pcluste commented on tests/integration-tests/tests/ad_integration/test_ad_integration.py:
```diff
@@ -155,7 +159,9 @@ def _store_secret(secret):
-    created_directory_stacks = {}
+    # TODO: do we need to worry about regions?
+    # TODO: protect modification of this dict via file locking
+    created_directory_stacks = defaultdict(dict)
```
Why the change is necessary

tilne replied:
the main reason i made the change was because it's now a nested dict with region and then stack type keys. This prevents having to check whether a dict has already been initialized in the region with something like:

```python
if region not in dict:
  dict[region] = dict()
```

it wasn't strictly necessary.

hanwen-pcluste replied:
Great. I wasn't aware of the `defaultdict`

hanwen-pcluste commented on tests/integration-tests/tests/ad_integration/cluster_user.py:
```diff
@@ -22,10 +24,10 @@ def __init__(self, user_num, test_datadir, cluster, scheduler, default_user_remo
         #       that creates users as part of the directory stack.
         self.password = "ApplesBananasCherries!"
-        self._validate_automatic_homedir_creation(cluster.head_node_ip)
+        self._validate_automatic_homedir_creation(self.cluster.head_node_ip)
```
Why the change is necessary? Same question for line 30

hanwen-pcluste commented on tests/integration-tests/tests/ad_integration/test_ad_integration.py:
```diff
@@ -164,78 +170,140 @@ def _check_ssm_success(ssm_client, command_id, instance_id):
+                )
+            else:
+                logging.info("Deleting %s stack named %s in region %s", stack_type, stack_name, region)
+                boto3.client("cloudformation", region_name=region).delete_stack(StackName=stack_name)
```
Sorry, this is a bug from me. It should've been
```
cfn_stacks_factory.delete_stack(stack_name, region)
```

tilne replied:
fixed.

Commit:
Explicitly set file permissions being tested

This change is motivated by the fact the default umask apparently
differs between AL2 and Ubuntu 20.04. In order to simplify the test,
explicitly set the permissions for all files whose permissions are
tested rather than depending on the default umask.

Signed-off-by: Tim Lane <tilne@amazon.com>
